Arrangement_on_surface_2/include/CGAL/draw_arrangement_2.h

```diff
@@ -0,0 +1,522 @@
+#if 0
+    if constexpr (std::experimental::is_detected_v<approximate_2_object_t, Gt>)
+    {
+      const auto* traits = this->m_arr.geometry_traits();
+      auto approx = traits->approximate_2_object();
+      draw_approximate_curve(curve, approx);
+      return;
+    }
+    draw_exact_curve(curve);
+#else
+    namespace bh = boost::hana;
+    auto has_approximate_2_object =
+      bh::is_valid([](auto&& x) -> decltype(x.approximate_2_object()){});
+    const auto* traits = this->m_arr.geometry_traits();
+    bh::if_(has_approximate_2_object(Gt{}),
+            [&](auto& /* x */) {
+              auto approx = traits->approximate_2_object();
+              auto has_operator = bh::is_valid(can_call_operator_curve<int>{});
+              bh::if_(has_operator(approx),
+                      [&](auto& x) { x.draw_approximate_curve(curve, approx); },
+                      [&](auto& x) { x.draw_exact_curve(curve); }
+                      )(*this);
+            },
+            [&](auto& x) { x.draw_exact_curve(curve); }
+            )(*this);
+#endif
+  }
+
+  /*! Add an approximation of a point.
+   * \param[in] p the (exact) point.
+   * Call this member function only if the geometry traits is equipped with
+   * the coordinate-approximation functionality of a point.
+   * This function must be inlined (e.g., a template) to enable the
+   * compiled-time dispatching in the function `draw_point()`.
+   */
+  template <typename Point, typename Approximate>
+  void draw_approximate_point(const Point& p, const Approximate& approx)
+  { this->add_point(approx(p)); }
+
+  //!
+  void draw_point(const Point& p) {
+    namespace bh = boost::hana;
+    auto has_approximate_2_object =
+      bh::is_valid([](auto&& x) -> decltype(x.approximate_2_object()){});
+    bh::if_(has_approximate_2_object(Gt{}),
+            [&](auto& x) {
+              const auto* traits = x.m_arr.geometry_traits();
+              auto approx = traits->approximate_2_object();
+              auto has_operator =
+                bh::is_valid([](auto&& x) -> decltype(x.operator()(Point{})){});
+              bh::if_(has_operator(approx),
+                      [&](auto& x) { x.draw_approximate_point(p, approx); },
+                      [&](auto& x) { x.add_point(p); }
+                      )(*this);
+            },
+            [&](auto& x) { x.add_point(p); }
+            )(*this);
+  }
+
+  //!
+  void add_ccb(Ccb_halfedge_const_circulator circ) {
+    auto curr = circ;
+    do {
+      auto new_face = curr->twin()->face();
+      if (m_visited.find(new_face) != m_visited.end()) continue;
+      m_visited[new_face] = true;
+      add_face(new_face);
+    } while (++curr != circ);
+  }
+
+  //!
+  void add_face(Face_const_handle face) {
+    for (auto it = face->inner_ccbs_begin(); it != face->inner_ccbs_end(); ++it)
+      add_ccb(*it);
+
+    for (auto it = face->outer_ccbs_begin(); it != face->outer_ccbs_end(); ++it)
+    {
+      add_ccb(*it);
+      draw_region(*it);
```

Comment:
MSVC is not happy with this conversion:
```
    90>C:\CGAL_ROOT\CGAL-5.6-Ic-90\include\CGAL/draw_arrangement_2.h(416): error C2664: 'void CGAL::Arr_2_basic_viewer_qt<GeometryTraits_2,Dcel>::add_ccb(CGAL::_HalfedgeDS_facet_const_circ<CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeometryTraits_2,Dcel>>::Halfedge,CGAL::I_Filtered_const_iterator<CGAL::internal::In_place_list_const_iterator<T,Alloc>,CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeometryTraits_2,Dcel>>::_Is_valid_halfedge,CGAL::internal::In_place_list_iterator<T,Alloc>,CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeometryTraits_2,Dcel>>::Halfedge,__int64,std::bidirectional_iterator_tag>,CGAL::Bidirectional_circulator_tag>)': cannot convert argument 1 from 'CGAL::Arr_halfedge<V,H,F> *const ' to 'CGAL::_HalfedgeDS_facet_const_circ<CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeomTraits,Dcel>>::Halfedge,CGAL::I_Filtered_const_iterator<CGAL::internal::In_place_list_const_iterator<T,Alloc>,CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeomTraits,Dcel>>::_Is_valid_halfedge,CGAL::internal::In_place_list_iterator<T,Alloc>,CGAL::Arrangement_on_surface_2<GeomTraits_,CGAL::Arr_bounded_planar_topology_traits_2<GeomTraits,Dcel>>::Halfedge,__int64,std::bidirectional_iterator_tag>,CGAL::Bidirectional_circulator_tag>' [C:\CGAL_ROOT\CGAL-5.6-Ic-90\cmake\platforms\MSVC2017-Debug-64bits\test\Arrangement_on_surface_2_Examples\circles.vcxproj]
```
Is there a better way to construct the CCB_const_halfedge_iterator?